src/AST.h

```diff
@@ -1,20 +1,22 @@
 #pragma once
-#include<string>
 
 #include <string>
+#include <memory.h>
+#include <boost/filesystem.hpp>
+namespace fs = boost::filesystem;
 
 #include <string>
 
 class Location {
 
 public:
 	Location(int firstLine, int firstCol, int lastLine, int lastCol,
-			const std::string &filename)
+			std::shared_ptr<fs::path> path)
 		: first_line(firstLine), first_col(firstCol), last_line(lastLine),
-		last_col(lastCol), filename(filename) {
+		last_col(lastCol), path(path) {
```

Comment:
Tiny optimization: `path(std::move(path))` - that saves us one extra reference count inc/dec

Reply:
Interesting idea - yeah that makes sense.
The constructor is its own context with no other code accessing path.
The caller of location already has to handle the shared pointer correctly and the location class as well.

Reply:
FYI: Effective Modern C++ has a discussion on this topic:
“Item 41: Consider pass by value for copyable parameters that are cheap to move and always copied.”